File: backend/app/schemas/ai_learning.py
from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class AIEventType(str, Enum):
    SUGGESTION_SHOWN = "suggestion_shown"
    ACCEPTED = "accepted"
    EDITED = "edited"
    REJECTED = "rejected"
    IGNORED = "ignored"
    FEEDBACK = "feedback"
    OUTCOME_OBSERVED = "outcome_observed"


class AIRecommendationStatus(str, Enum):
    PENDING = "pending"
    ACCEPTED = "accepted"
    DISMISSED = "dismissed"
    STALE = "stale"


class AIRecommendationPriority(str, Enum):
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"
    INFO = "info"


class AICorrectionCreate(BaseModel):
    field_path: str = Field(..., min_length=1, max_length=255)
    proposed_value: Optional[Any] = None
    final_value: Optional[Any] = None
    correction_reason: Optional[str] = None
    confidence_score: Optional[float] = Field(None, ge=0, le=1)


class AICorrectionResponse(BaseModel):
    id: int
    event_id: Optional[int] = None
    recommendation_id: Optional[int] = None
    source_module: str
    entity_type: Optional[str] = None
    entity_id: Optional[int] = None
    field_path: str
    proposed_value: Optional[Any] = None
    final_value: Optional[Any] = None
    correction_reason: Optional[str] = None
    confidence_score: Optional[float] = None
    created_by: Optional[int] = None
    created_at: datetime

    class Config:
        from_attributes = True


class AIInteractionEventCreate(BaseModel):
    event_type: AIEventType
    source_module: str = Field(..., min_length=1, max_length=80)
    ai_feature: Optional[str] = Field(None, max_length=120)
    surface: Optional[str] = Field(None, max_length=120)
    entity_type: Optional[str] = Field(None, max_length=80)
    entity_id: Optional[int] = None
    recommendation_id: Optional[int] = None
    context_summary: Optional[str] = Field(None, max_length=4000)
    event_payload: Optional[Dict[str, Any]] = None
    confidence_score: Optional[float] = Field(None, ge=0, le=1)
    prompt_version: Optional[str] = Field(None, max_length=120)
    model_version: Optional[str] = Field(None, max_length=120)
    corrections: List[AICorrectionCreate] = Field(default_factory=list)


class AIInteractionEventResponse(BaseModel):
    id: int
    event_type: str
    source_module: str
    ai_feature: Optional[str] = None
    surface: Optional[str] = None
    entity_type: Optional[str] = None
    entity_id: Optional[int] = None
    recommendation_id: Optional[int] = None
    context_summary: Optional[str] = None
    event_payload: Dict[str, Any] = Field(default_factory=dict)
    confidence_score: Optional[float] = None
    prompt_version: Optional[str] = None
    model_version: Optional[str] = None
    created_by: Optional[int] = None
    created_at: datetime
    corrections: List[AICorrectionResponse] = Field(default_factory=list)

    class Config:
        from_attributes = True


class AIRecommendationCreate(BaseModel):
    source_module: str = Field(..., min_length=1, max_length=80)
    recommendation_type: str = Field(..., min_length=1, max_length=80)
    priority: AIRecommendationPriority = AIRecommendationPriority.MEDIUM
    title: str = Field(..., min_length=1, max_length=255)
    summary: str = Field(..., min_length=1)
    rationale: Optional[str] = None
    target_entity_type: Optional[str] = Field(None, max_length=80)
    target_entity_id: Optional[int] = None
    suggested_action: Optional[Dict[str, Any]] = None
    evidence: Optional[List[Dict[str, Any]]] = None
    impact: Optional[Dict[str, Any]] = None
    confidence_score: float = Field(0.5, ge=0, le=1)
    prompt_version: Optional[str] = Field(None, max_length=120)
    model_version: Optional[str] = Field(None, max_length=120)
    expires_at: Optional[datetime] = None


class AIRecommendationResponse(BaseModel):
    id: int
    source_module: str
    recommendation_type: str
    status: str
    priority: str
    title: str
    summary: str
    rationale: Optional[str] = None
    target_entity_type: Optional[str] = None
    target_entity_id: Optional[int] = None
    suggested_action: Dict[str, Any] = Field(default_factory=dict)
    evidence: List[Dict[str, Any]] = Field(default_factory=list)
    impact: Dict[str, Any] = Field(default_factory=dict)
    confidence_score: float
    prompt_version: Optional[str] = None
    model_version: Optional[str] = None
    status_reason: Optional[str] = None
    created_by: Optional[int] = None
    accepted_by: Optional[int] = None
    dismissed_by: Optional[int] = None
    created_at: datetime
    updated_at: datetime
    acted_at: Optional[datetime] = None
    expires_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class AIRecommendationActionRequest(BaseModel):
    reason: Optional[str] = Field(None, max_length=2000)


class AIRecommendationFeedbackRequest(BaseModel):
    feedback: str = Field(..., min_length=1, max_length=4000)
    rating: Optional[int] = Field(None, ge=1, le=5)
    event_payload: Dict[str, Any] = Field(default_factory=dict)


class AIOutcomeCreate(BaseModel):
    recommendation_id: Optional[int] = None
    source_module: str = Field(..., min_length=1, max_length=80)
    outcome_type: str = Field(..., min_length=1, max_length=80)
    entity_type: Optional[str] = Field(None, max_length=80)
    entity_id: Optional[int] = None
    metric_name: Optional[str] = Field(None, max_length=120)
    metric_value: Optional[float] = None
    baseline_value: Optional[float] = None
    target_value: Optional[float] = None
    outcome_payload: Optional[Dict[str, Any]] = None
    observed_at: Optional[datetime] = None


class AIOutcomeResponse(BaseModel):
    id: int
    recommendation_id: Optional[int] = None
    source_module: str
    outcome_type: str
    entity_type: Optional[str] = None
    entity_id: Optional[int] = None
    metric_name: Optional[str] = None
    metric_value: Optional[float] = None
    baseline_value: Optional[float] = None
    target_value: Optional[float] = None
    outcome_payload: Dict[str, Any] = Field(default_factory=dict)
    observed_at: datetime
    created_by: Optional[int] = None
    created_at: datetime

    class Config:
        from_attributes = True


class AIAggregationSummary(BaseModel):
    companies_processed: int
    recommendations_created: int
    stale_recommendations: int

File: frontend/src/types/aiLearning.ts
export type AIEventType =
  | 'suggestion_shown'
  | 'accepted'
  | 'edited'
  | 'rejected'
  | 'ignored'
  | 'feedback'
  | 'outcome_observed';

export type AIRecommendationStatus = 'pending' | 'accepted' | 'dismissed' | 'stale' | 'snoozed';
export type AIRecommendationPriority = 'high' | 'medium' | 'low' | 'info';

export interface AICorrectionInput {
  field_path: string;
  proposed_value?: unknown;
  final_value?: unknown;
  correction_reason?: string;
  confidence_score?: number;
}

export interface AIInteractionEventInput {
  event_type: AIEventType;
  source_module: string;
  ai_feature?: string;
  surface?: string;
  entity_type?: string;
  entity_id?: number;
  recommendation_id?: number;
  context_summary?: string;
  event_payload?: Record<string, unknown>;
  confidence_score?: number;
  prompt_version?: string;
  model_version?: string;
  corrections?: AICorrectionInput[];
}

export interface AIRecommendation {
  id: number;
  source_module: string;
  recommendation_type: string;
  status: AIRecommendationStatus;
  priority: AIRecommendationPriority;
  title: string;
  summary: string;
  rationale?: string;
  target_entity_type?: string;
  target_entity_id?: number;
  suggested_action?: Record<string, unknown>;
  evidence?: Array<Record<string, unknown>>;
  impact?: Record<string, unknown>;
  confidence_score: number;
  prompt_version?: string;
  model_version?: string;
  status_reason?: string;
  created_by?: number;
  accepted_by?: number;
  dismissed_by?: number;
  created_at: string;
  updated_at: string;
  acted_at?: string;
  expires_at?: string;
  /** Deterministic Action Inbox ranking score; only populated by the list endpoint. */
  score?: number | null;
}

export interface AIRecommendationApplyResult {
  recommendation: AIRecommendation;
  applied: boolean;
  apply_result?: Record<string, unknown> | null;
  apply_error?: string | null;
}

/** True when Accept can run an allowlisted ERP mutation. */
export function recommendationIsApplyable(rec: AIRecommendation): boolean {
  const action = rec.suggested_action || {};
  const autonomy = String(action.autonomy || 'suggest_only');
  const type = String(action.type || '');
  if (!type) return false;
  return (
    autonomy === 'apply_on_accept' ||
    autonomy === 'execute_controlled' ||
    autonomy === 'auto_execute'
  );
}

export interface AIRecommendationInput {
  source_module: string;
  recommendation_type: string;
  priority?: AIRecommendationPriority;
  title: string;
  summary: string;
  rationale?: string;
  target_entity_type?: string;
  target_entity_id?: number;
  suggested_action?: Record<string, unknown>;
  evidence?: Array<Record<string, unknown>>;
  impact?: Record<string, unknown>;
  confidence_score?: number;
  prompt_version?: string;
  model_version?: string;
  expires_at?: string;
}

export interface AIRecommendationFeedbackInput {
  feedback: string;
  rating?: number;
  event_payload?: Record<string, unknown>;
}

export interface AIOutcomeInput {
  recommendation_id?: number;
  source_module: string;
  outcome_type: string;
  entity_type?: string;
  entity_id?: number;
  metric_name?: string;
  metric_value?: number;
  baseline_value?: number;
  target_value?: number;
  outcome_payload?: Record<string, unknown>;
  observed_at?: string;
}
